proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
 All cross-chain messages must have the correct format, which is checked by the following logic:
 
 ```python
-validateFormat(ccm):
-    if size(ccm) > MAX_CCM_SIZE bytes:
-        return False
-    if ccm does not follow crossChainMessageSchema:
-        return False
-    return True
+def validateFormat(ccm: CCM) -> None:
+    # Try to encode ccm.
+    # This raises an error if the ccm does not follow the correct schema.
+    ccmBytes = encode(crossChainMessageSchema, ccm)
+    # Check ccm size.
+    if ccmBytes > MAX_CCM_SIZE:
+        raise Exception(f"Cross-chain message size is larger than {MAX_CCM_SIZE}.")
 ```
 
 #### apply
 
 The `apply` function applies a cross-chain message, in the context of a cross-chain update `ccu`.
 
 ```python
-# The ccu including this message is passed as context to this function
-apply(ccu, ccm):
-    if terminatedStateAccount(ccm.sendingChainID) exists:
-        ccm is discarded and has no further effect
-
-    # Call the beforeApplyCCM functions from other modules
-    # in particular, this will assign the message fee to the ccu sender
-    for each module mdl for which beforeApplyCCM exists:
-        mdl.beforeApplyCCM(ccu, ccm)
-        if the previous call fails or returns False:
+def apply(ccu: CCU, ccm: CCM) -> None:
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+    # Apply the message if the logic is supported
+    if (ccm.moduleID, ccm.crossChainCommandID) is supported: 
+        # Create a state snapshot.
+
+        snapshot = snapshot of the current state
+        try: 
+            # Call the beforeCrossChainCommandExecution functions from other modules
+            # For example, the Token module assigns the message fee to the CCU sender.
+            for each module mdl for which beforeCrossChainCommandExecution exists:
+                mdl.beforeCrossChainCommandExecution(ccu, ccm)
```

Comment:
Here the whole CCU is available in these hooks. I assume this is mainly as the Token module needs the CCU sender. I am not sure how this is handled in the current implementation, but I think it would be best to minimize what is accessible in this hook so the modules do not do something that relies on the specific CCU structure. These minimal properties could then be accessible in the CCM processing context similar to the transaction processing context.

Reply:
I see your point, but I am not sure if we can decide what this minimal interface should be. Also, would it be such a big problem if the whole CCU is accesible?

Reply:
I would suggest to only make the base transaction object, but not the params of the CCU accessible. Minimizing what is accessible makes the architecture generally more robust as then an update to the CCU params schema cannot affect modules.

Maybe we could ask the devs about their opinion on the topic.

Reply:
sounds good to ask the devs. In the past we have been quite flexible on this in the LIP specs, but it's good to be precise here

Reply:
@shuse2 @ishantiw opinions?

Reply:
I think we don't have to settle this as part of the review. Maybe you could add an agenda point to our meeting and create a GitHub discussion for it. Then we can move forward with this review.